cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/AggregatingFunctionsTest.scala

```diff
@@ -386,5 +394,71 @@ class AggregatingFunctionsTest extends DocumentingTest {
         resultTable()
       }
     }
+    section("Grouping keys", "grouping-keys") {
+      p("""The expressions which does not contain any aggregation function are the grouping keys of the aggregation.
+          #If there are no grouping keys, the aggregation will be over all rows.
+          #For expressions that contain an aggregation function, the leaves of the expressions must be one of:
```

Comment:
Is there some way to explain the reasoning here, perhaps by grouping the alternatives like

> For expressions that contain an aggregation function, the leaves of the expressions must each be one of:
> * An aggregation
> * A constant, i.e.
>   * A literal
>   * A Parameter
> * A simple grouping key, i.e.
>   * A variable - if that variable is a projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`)
>   * Property access - ONLY IF that property access is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`)
>   * Map access - ONLY IF that map access is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)
> * A local variable, i.e.
>   * A variable introduced in the expression (e.g the `x` in `RETURN n, n.prop + size([ x IN range(1, 10) | x ])`)

Would this help users grasp the reasoning better? I'm not sure, I'll leave it up to you :smile:

Reply:
I would not like that document in a manual. The manual describes the state of the product and how to use it. It does not given technical design background. If there is the desire to provide technical design background for the _really_ interested user, there are other formate, e.g. developer blog posts, where that can be done.

BTW: "the leaves of the expressions" feels also too language-nerdy for the manual.

We can provide some reasoning by staging the explanation a bit. Sketch:

=====

Expressions that contain an aggregation function are aggregation expressions. Simple aggregation expression only an aggregation function an its argument. For instance, `SUM(x.a)` is an aggregation expressions that only consists of the aggregation function `SUM( )` with `x.a` as it argument. Aggregation expression are also allowed to be more complex, where the result of one or more aggregation functions are operands to other computation. For instance, `0.1 * (SUM(x.a) / COUNT(x.b))` is an aggregation expressions that contains two aggregation functions, `SUM( )` with `x.a` as it argument and `COUNT( )` with `x.b` as it argument. Both are operand to a numeric division. The result of this numeric division is an operand to a numeric multiplication. The numeric multiplication has the value `0.1` as its second operand.

For aggregation expressions to be correctly computable for of the buckets formed by the grouping key, they have to fulfill some requirements. Specifically, each operand in an aggregation expression has to be either

- an aggregation function, e.g. `SUM(x.a)`,
- a constant, e.g. `0.1`,
- a parameter, e.g. `$param`,
- a grouping key, e.g. ....
- a local variable, e.g. ..., or
- a subexpression, all whose operands are operands allowed in an aggregation expression.

// pro example

// counter example

Grouping keys themselves can be complex expression. For better query readability, Cypher only recognizes operands in aggregation expressions to be grouping key if the grouping key is either

- A variable - ...
- A property access - ...
- A map access - ...

If more complex grouping key are need as operands in aggregation expression, it is always possible to project them in advance with `WITH`.

// example (hopefully implicitly demonstrates, that the extra project steps help query readability.)

=====

Again this a sketch. No ready-to-be-used text snippet. It is meant as a suggest of how the text could be made more digestible in a way that is also suitable for a manual – at least according to my uneducated understanding of manuals. The idea of the sketch is to stage the explanation a bit more and provide some high level reasons without going into details, e.g. "to be correctly computable", "for better query readability". It certainly needs more editorial love and integration with the other bits. The more concrete the examples the better. Feel free to improve in any way you think suitable.